crates/sql/src/ast/subscription.rs

```diff
@@ -0,0 +1,258 @@
+//!
+//! - [`SELECT`](SqlProject)
+//! - [`FROM`](SqlFrom)
+//! - [`JOIN`](SqlJoin)
+//! - [`ON`](SqlOn)
+//! - [`WHERE`](SqlSelection)
+//!
+//! ### Subscription Grammar
+//!
+//! **WARNING:** This grammar does not guarantee the semantic validity of its queries, only their syntactic correctness.
+//!
+//! ```ebnf
+//! query
+//!     = SELECT projection FROM relation [ WHERE predicate ]
+//!     ;
+//!
+//! projection
+//!     = STAR
+//!     | ident '.' STAR
+//!     ;
+//!
+//! relation
+//!     = table [ [AS] ident ] [ [INNER] JOIN table [ [AS] ident ] ON predicate_on ]
+//!     ;
+//!
+//! columnExpr
+//!     = column
+//!     | qualfied
+//!     ;
+//!
+//! predicate
+//!     =  expr
+//!     | predicate AND predicate
+//!     ;
+//!
+//! predicate_on
+//!     =  columnExpr '=' columnExpr
+//!     | predicate_on AND predicate_on
+//!     ;
+//!
+//! expr
+//!     = literal
+//!     | ident
+//!     | qualified
+//!     | expr op expr
+//!     ;
+//!
+//! qualified
+//!     = ident '.' ident
+//!     ;
+//!
+//! op
+//!     = '='
+//!     | '<'
+//!     | '>'
+//!     | '<' '='
+//!     | '>' '='
+//!     ;
+//!
+//! literal
+//!     = INTEGER
+//!     | STRING
+//!     | HEX
+//!     | TRUE
+//!     | FALSE
+//!     ;
+//! ```
+//!
+use crate::ast::*;
+use crate::errors::{SqlError, SqlUnsupported};
+
+/// The maximum number of `JOIN` clauses allowed in a `SQL` subscription query.
+///
+/// This is a limitation to avoid complex queries that could be hard to optimize.
+pub const MAX_JOINS: usize = 2;
+
+/// The comparison operators supported in `SQL` subscription queries.
+///
+/// ```ebnf
+/// op
+///     = '='
+///     | '<'
+///     | '>'
+///     | '<' '='
+///     | '>' '='
+///     ;
+/// ```
+#[derive(Debug, Copy, Clone, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub enum SqlOpCmp {
+    Eq,
+    Lt,
+    LtEq,
+    Gt,
+    GtEq,
+}
+
+impl TryInto<SqlOpCmp> for OpCmp {
+    type Error = SqlError;
+
+    fn try_into(self) -> Result<SqlOpCmp, Self::Error> {
+        match self {
+            OpCmp::Eq => Ok(SqlOpCmp::Eq),
+            OpCmp::Lt => Ok(SqlOpCmp::Lt),
+            OpCmp::LtEq => Ok(SqlOpCmp::LtEq),
+            OpCmp::Gt => Ok(SqlOpCmp::Gt),
+            OpCmp::GtEq => Ok(SqlOpCmp::GtEq),
+            _ => Err(SqlUnsupported::SubscriptionNotEq.into()),
+        }
+    }
+}
+
+/// The `SELECT` clause for `SQL` subscription queries.
+///
+/// ```ebnf
+/// SELECT ( '*' | table '.' '*' )
+/// ```
+#[derive(Debug, Eq, PartialEq)]
+pub enum SqlProject {
+    /// An qualified `table.*`
+    QualifiedWildcard { table: Identifier },
+    /// An unqualified `SELECT *`
+    Wildcard,
+}
+
+/// The `ON` clause of `JOIN` for `SQL` subscription queries.
+///
+/// ```ebnf
+/// predicate_on
+///    = columnExpr '=' columnExpr
+///   | predicate_on AND predicate_on
+///  ;
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SqlOn {
+    And { lhs: Box<SqlOn>, rhs: Box<SqlOn> },
+    Eq { lhs: SqlColumn, rhs: SqlColumn },
+}
```

Comment:
We could be permissive, allow any expression here, and then catch invalid ones during semantic analysis. I think that's probably what we should do.